### DISCUSSION TITLE
Why are there no NFT-s?

### DISCUSSION THREAD

**Discussion by razgraf** [sablier-labs/flow#34] [General]
I've noticed that Flow does not have any NFTs. For reference (and possibly debate), would you mind describing the problem and the reasoning behind this decision @andreivladbrg?

**Comment by andreivladbrg** [comment]
Yes, this is correct, we do not have the NFTs in open-ended contracts (at least not anymore, initially they were implemented).

The rational behind is: 
- the product is oriented towards the sender, and the NFT would be minted for the recipient.
- there is no "non-cancelable" optionality (all streams are cancelable).
- no deposit is required, thus there is no exact value of the NFT (it could actually entail "debt").
- the sender is able to retrieve funds from an ongoing stream (but not the amount that has already been streamed).
- there is no fixed duration for the stream.
- it would increase the gas fees and complexity. I started building the open-ended contracts with the goal of being as minimalist as possible, since the main use-case for them is straightforward (salaries with top-ups).
- we would need a new design for the NFT

Having said these points, I consider that there is no value in adding NFTs to the open-ended product, other than to say, "Look, we have NFTs for the payroll contracts, but LlamaPay doesn't." 

Also, this NFTs discussion was initially raised by PRB in the first presentation of the contracts [here](https://github.com/sablier-labs/v2-open-ended/discussions/1) (Spec Questions) and I've answered it [this discussion](https://github.com/sablier-labs/company-discussions/discussions/6#discussioncomment-7304440) and no counter-argument was presented.

**Reply by andreivladbrg** [reply]
Also, one comment from a private discussion with @smol-ninja, which I agree with:

"
I think its better if only sender is allowed to change recipient. The most [anticipated use case](https://sablier.slack.com/archives/C032HBFV4QJ/p1712570093176659) seems to be payroll and from the legal point of view, it might not be a good idea to allow changing stream ownership by the recipient. By introducing ERC-721 also means we will have to introduce a new field similar to isTransferable
"

**Later edit:** we did not take any decision on this, we were just brainstorming the idea

**Reply by smol-ninja** [reply]
To correct the quoted comment, we did make a decision on this. We decided to open a discussion on whether open-ended streams should have NFT or not since in the current design, there is no concept of NFT. 

There was one point we also discussed which is that Llamapay is a simple contract which makes it extremely cheap to use and adding NFT increases gas, however, I don't see that as a strong reason because gas is not the only reason why users use protocols (otherwise Llamapay would have been no. 1). Sablier has a great design, has never been hacked i.e. safest protocol to use, and a dedicated team constantly working on this niche.

I do agree that adding NFT means we will also have to design specs over the transferability of the stream similar to lockups. Because due to compliance reasons, some companies may not want their payee to be able to change the recipient address.

**Comment by razgraf** [comment]
NFTs are actually a pretty good growth hack, as well as a means to transfer value from one wallet to another.

Having NFTs would make the stream a money-lego, with users being able to add it to lending markets, borrow against it etc. One of the biggest use-cases with payroll is [payday loans](https://www.citizensadvice.org.uk/debt-and-money/borrowing-money/types-of-borrowing/loans/payday-loans/taking-out-a-payday-loan/), so having the freedom to move the stream is required.

The problem of quantifying or classifying a stream is a layer that 3rd parties can asses, just like in the case of cancelable/non-cancelable streams. Open-ended streams can still accrue real value inside (e.g. by not withdrawing), represent a promise of future income etc. - by using the transferability aspect of the NFT, we'd open it to the same ecosystem of opportunities the lockup one have.

On the "Look, we have NFTs for the payroll contracts, but LlamaPay doesn't.", that's somewhat true, but as I said: it's a growth hack tool too. I'm worried open-ended would end up looking like a rip-off of Llama Pay's product if we don't engineer new functionalities in it or innovate in any way - the gas costs are negligible, as we're not in the market of offering the no. 1 most efficient product, but in the market of offering the no. 1 most complete and beautiful product (from design to code, from concept to functionality).


> I've answered it https://github.com/sablier-labs/company-discussions/discussions/6#discussioncomment-7304440 and no counter-argument was presented.

Sorry about that, I haven't had a lot of time to dig into the existing discussions for OE yet, so some of these points I'm seeing for the first time. Feel free to ping me in them if I can give any sort of input - will also make an effort to read them one by one soon and ask questions.

**Reply by andreivladbrg** [reply]
> NFTs are actually a pretty good growth hack

I can agree with this

> Having NFTs would make the stream a money-lego, with users being able to add it to lending markets, borrow against it etc. One of the biggest use-cases with payroll is [payday loans](https://www.citizensadvice.org.uk/debt-and-money/borrowing-money/types-of-borrowing/loans/payday-loans/taking-out-a-payday-loan/), so having the freedom to move the stream is required.

this is the problem - I don't see this working in an "open" environment given the points I've mentioned (too many "variables" that someone should take account for - on the other hand for lockup the NFT makes total sense)

> I'm worried open-ended would end up looking like a rip-off of Llama Pay's product if we don't engineer new functionalities in it or innovate in any way - 

It is a real concern that we should have. In terms of contracts, the way it is implemented is quite different (@smol-ninja, please correct me if the implementation is not different). We focus on the Stream entity, whereas LlamaPay focuses on the payer details.

And, for LlamaPay, you can't deposit for specific recipients, meaning you can't prioritize. I plan to open a discussion outlining all the differences.

> the gas costs are negligible, as we're not in the market of offering the no. 1 most efficient product, but in the market of offering the no. 1 most complete and beautiful product (from design to code, from concept to functionality).

Agree with this

> Sorry about that, I haven't had a lot of time to dig into the existing discussions for OE yet

No prob

I've just restarted working for OE, so I need to open multiple discussions about the directions and what extra functionalities we should have. As a start, the [README](https://github.com/sablier-labs/v2-open-ended/blob/main/README.md) should be a good reference.

**Reply by smol-ninja** [reply]
> this is the problem - I don't see this working in an "open" environment given the points I've mentioned

Can you please explain it again? I didn't get what you meant here?

> please correct me if the implementation is not different

Yes its different in the context of that each stream on Llamapay is a separate contract while in OE, its just one contract that handles all the streams. This opens up for new functionalities such as making deposits to multiple streams at once (this would actually be cheaper in our case).

I agree with @razgraf comment and I think we should include NFT in OE too.l

**Reply by andreivladbrg** [reply]
> Can you please explain it again? I didn't get what you meant here?

When you want to borrow against something, in general, there are some criteria that should be respected. In the context of OE streams, they would be considered too "volatile"

there are no `non-cancelable` streams - sender can cancel the stream without even depositing to it 
the `withdraw` function is public - anyone can trigger it 
not a fixed duration with a fixed rate - you can't say that by the end of X date I will have this amount

Also, in the context of lockup with a non-cancelable streams, there isn't a stable integration made atm (This is why I am pessimistic about this working for open-ended or having real value). We still need to build our own NFT lending platform. 

> Yes its different in the context of that each stream on Llamapay is a separate contract while in OE, its just one contract that handles all the streams. This opens up for new functionalities such as making deposits to multiple streams at once (this would actually be cheaper in our case).

Yeah, besides the factory pattern design, they don't have data structure for a "Stream"

**Reply by smol-ninja** [reply]
I got it now. I understand your point about the risks involved in lending with OE NFTs and I agree with it. However, there are two other features that Sablier NFTs offer (and are true in the present):

1. **Visualization**: The NFT provides a nice visualization of the stream, displaying all the relevant information. Even without lending (which is anyways not an active utility at the moment), the NFT itself represents ownership of the stream. Users can see it in their crypto wallets and have the assurance that the stream belongs to them. Without the NFT, the payment becomes invisible and can only be verified on the Sablier user interface.


2. **Transferability**: While, to me, it makes sense to only allow the Sender to change the recipient. From the perspective of the recipients, it's helpful to have the option to transfer the stream. For example, if their wallet gets compromised, recipients can take quick action (as opposed to requesting Sender to change it) as hackers can only stream ERC20 tokens. Additionally, keeping the specs similar to lockup contracts ensures easy onboarding for our current users.

**Comment by smol-ninja** [comment]
Do we all agree to have NFT for OE? @andreivladbrg tagging you for your vote.

**Comment by andreivladbrg** [comment]
On the last all-hands call, we have decided to implement the NFT in OE as well. However, I want to mention that doing so would considerably increase the complexity and the gas costs for the `create` `adjustRatePerSecond` and `withdraw` functions if we choose to keep the `adjustRatePerSecond` function.

#### Why

Based on Rusty's [findings](https://github.com/cantinasec/review-sablier/issues/11) from the first audit, if the NFT is transferred to a blacklisted address, the cancel function would revert. To address this, we would need to implement a new function, `withdrawWhenCanceled`, which withdraws the balance at that moment in time. We can't apply the methodology from the lockup contract here because we do not have a fixed deposit amount.

In the case of `adjustRatePerSecond`, we need to perform the `ERC20` transfer to the recipient because of the same [reason](https://github.com/sablier-labs/v2-open-ended?tab=readme-ov-file#how-it-works) explained in the `README` on why we need to update the `lastTimeUpdate` variable:

>>because it would cause loss of funds for the recipient if the previous rate was higher or gain of funds if
    the previous rate was lower

If we can no longer transfer funds (for the blacklist reason) we would need to add a new amount (`beforeAdjustAmount` - or any other name) in the Stream struct, which will be summed up in the `withdrawableAmount` function

---

Since, IMO the best argument for using an NFT is its visual aspect, what would you say if we do not inherit from `ERC721` but instead only implement a `tokenURI` function ? In our case it would be `streamURI`😅 - might not be compatible with platforms like opensea
This approach would provide something visual without adhering to ERC721 (no transfers).

**Reply by andreivladbrg** [reply]
> withdrawWhenCanceled, which withdraws the balance at that moment in time. We can't apply the methodology from the lockup contract here because we do not have a fixed deposit amount.

just realized that this is a vector attack, since all canceled streams can be restarted - what happens if the recipient does not withdraw the assets after cancel, and then the stream is restarted? 🤔

**Reply by smol-ninja** [reply]
Curious what is the problem with `beforeAdjustAmount`? Lets call it: `uint128 withdrawableAmount` (in slot 3) field `Stream` struct. This will store the amount that can be withdrawn at any point of time.

As a result of this, we can make the following change:
-  `_cancel()` only cancels and refunds to sender and let recipient / any unknown address trigger withdraw separately. It also updates `withdrawableAmount`.
- `restartStream()` does not update `withdrawableAmount` but update `lastTimeUpdate`
- `withdraw` calculates withdrawable amount using `withdrawableAmount` and `lastTimeUpdate`. It also sets `withdrawableAmount` to zero.
- `refundFromStream` updates `withdrawableAmount` and `lastTimeUpdate` but does not withdraw funds to recipients.

This solves the following problems:
- vector attack that you mentioned
- Rusty's https://github.com/cantinasec/review-sablier/issues/11 from the first audit. Now the `cancel` would not revert. So no need for `withdrawWhenCanceled`.

> Since, IMO the best argument for using an NFT is its visual aspect

Thats one advantage. Another advantages are `Transferability` and `Ownership`. Being able to prove that you own a stream is, imo, a bigger advantage of representing aa stream with an NFT. Its the same argument as DEXes like Uniswap could represent liquidity without introducing ERC-20 LP tokens.

**Reply by andreivladbrg** [reply]
thanks shub for the proposed change. there are some edge cases that are not covered

> (in slot 3) field Stream struct

it will be added in the slot 4 since the address (160 bits) + uint128 (128 bits) equals 288, which is more than 32 bytes
and it means that the `create` function will cost 21,000 gas more 

> uint128 withdrawableAmount

i think the name is misleading, what about something like `remainingAmount`? (there should be better names)

currently, the withdrawable amount is: `min(streamedAmount, balance)`, if we add this variable it will be changed to: `min(streamedAmount + newAmountAdded, balance)`?
what if the `newAmountAdded` is greater than `balance`? you should be able to withdraw this amount, but since the balance is less, would it revert?

what happens if you want to withdraw an amount less than newAmountAdded?

> _cancel() only cancels and refunds to sender and let recipient / any unknown address trigger withdraw separately. It also updates withdrawableAmount

it means that we need to make the `withdraw` function callable when the stream is canceled. also, the `rps` must be set to 0.

> withdraw calculates withdrawable amount using withdrawableAmount and lastTimeUpdate. It also sets withdrawableAmount to zero.

will it always be transferred the new amount in full if it is greater than 0? does the recipient have the option to withdraw only half? (imo, it should be the full amount > 0)

what time variable do you pass in the `withdraw` function?

> refundFromStream updates withdrawableAmount and lastTimeUpdate but does not withdraw funds to recipients

would it update the `lastTimeUpdate` to `block.timestamp` and the new amount added in storage will be updated with the full streamed amount?

and lastly, the new amount and lastTimeUpdate need to be updated in  `adjustRatePerSecond` function

----

> Thats one advantage. Another advantages are Transferability and Ownership. Being able to prove that you own a stream is, imo, a bigger advantage of representing aa stream with an NFT. Its the same argument as DEXes like Uniswap could represent liquidity without introducing ERC-20 LP tokens.

yeah, i understand the point, but as i said before, i am pessimistic about making it work (if we don't have something stable for the lockup version)

**Reply by smol-ninja** [reply]
> it will be added in the slot 4 since the address (160 bits) + uint128 (128 bits) equals 288

Yes I agree. It will require a new slot. However, another thought, recipient can also be calculated as the owner of NFT. Should we replace `recipient` from slot 1 with `withdrawableAmount` in that case, ofc only if its an NFT? 

> it means that the create function will cost 21,000 gas more

Why? When a stream is created `remainingAmount` will not get updated as its still zero at the time of creation so I think it won't add gas to `create`.

> what about something like `remainingAmount`

I am fine with that too. Since `remainingAmount` stores amount that can be withdrawn at any given point of time (assuming sufficient balance), so with this, `withdrawableAmountOf` can be calculated as `min(balance, remainingAmount)`, right?

> but since the balance is less, would it revert?

I didn't get this. Could you please explain again? What is `newAmountAdded`?

>  means that we need to make the withdraw function callable when the stream is canceled. also, the rps must be set to 0

Yes. `withdraw` can be called even after it has been cancelled. Since `cancel` would not transfer funds to the recipient anymore. Another advantage of this is that if the recipients get blocked by stablecoins such as USDT / USDC (may be because they exploited another protocol), sender would be able to `cancel` and refund his remaining amount. While in the current design, sender funds can get locked.

> will it always be transferred the new amount in full if it is greater than 0? does the recipient have the option to withdraw only half? (imo, it should be the full amount > 0)

It depends on the `time` variable passed in `withdraw` function. if its `block.timestamp` it should be min(remainingAmount, balance).

> what time variable do you pass in the withdraw function?

Its an input variable.

> would it update the lastTimeUpdate to block.timestamp and the new amount added in storage will be updated with the full streamed amount?

Yes in case of `refundfromStream`.

---------------
1. Why are you pessimistic about making it work?
2. if we don't have something stable for the lockup version - what do you mean by this?

**Reply by smol-ninja** [reply]
~@andreivladbrg, should I proceed with the implementation to directly measure and compare the actual gas differences? It would also help us identify the hidden gaps in the implementation. If the results are better than our initial assumptions, we can use that as a solid basis to conclude this discussion.~ We both have agreed that we would have NFT implementation.

**Reply by andreivladbrg** [reply]
> What is newAmountAdded?

i will explain this first

`newAmountAdded = remainingAmount`

i've said: "there should be better names"

so i didn't want to refer to it as `remainingAmount` and i used `newAmountAdded` in all my following sentences, but now, since you said that this name works, we can refer to it as `remainingAmount`


> Should we replace recipient from slot 1 with withdrawableAmount in that case, ofc only if its an NFT?
When a stream is created remainingAmount will not get updated as its still zero at the time of creation so I think it won't add gas to create

yes, we will remove the `recipient` from `Stream` struct, but it would still be placed in storage 

it doesn't matter if the `remainingAmount` is zero, it would still require a new slot in the Stream struct, since we can't pack together all the other variable

i've also tested this in remix, see code below:

<details><summary>Test code</summary>

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity =0.8.19;

struct StreamNoAmount {
    // slot 0
    uint128 balance;
    uint128 ratePerSecond;
    // slot 1
    address sender;
    uint40 lastTimeUpdate;
    bool isStream;
    bool isCanceled;
    uint8 assetDecimals;
    // slot 2
    address asset;
}

struct StreamWithAmount {
    // slot 0
    uint128 balance;
    uint128 ratePerSecond;
    // slot 1
    uint128 remainingAmount;
    uint40 lastTimeUpdate;
    bool isStream;
    bool isCanceled;
    // slot 2
    address asset;
    uint8 assetDecimals;
    // slot 3
    address sender;
}

contract Test_NoAmount {
    // slots 0, 1, 2
    mapping(uint256 => StreamNoAmount) internal _streamsNoAmount;

    // slot 3
    mapping(uint256 => address) public owners;

    // slot 4
    uint256 public index = 1;

    // gas cost: 134,909
    function test_StreamNoAmount() public {
        _streamsNoAmount[index] = StreamNoAmount({
            asset: address(0x1234),
            assetDecimals: 18,
            balance: 0,
            isStream: true,
            isCanceled: false,
            lastTimeUpdate: uint40(block.timestamp),
            ratePerSecond: 0.01e18,
            sender: address(0xCAFE)
        });

        owners[index] = address(0xBEEF);
        ++index;
    }
}

contract Test_WithAmount {
    // slots 0, 1, 2, 3
    mapping(uint256 => StreamWithAmount) internal _streamsWithAmount;

    // slot 4
    mapping(uint256 => address) public owners;

    // slot 5
    uint256 public index = 1;

    // gas cost: 160,471
    function test_StreamWithAmount() public {
        _streamsWithAmount[index] = StreamWithAmount({
            asset: address(0x1234),
            assetDecimals: 18,
            balance: 0,
            isStream: true,
            isCanceled: false,
            lastTimeUpdate: uint40(block.timestamp),
            ratePerSecond: 0.01e18,
            remainingAmount: 0,
            sender: address(0xCAFE)
        });

        owners[index] = address(0xBEEF);
        ++index;
    }
}

```

</details> 

what i've not mentioned is that it would also increase the gas costs of the function that changes the value of `remainingAmount` (e.g. `cancel`)

> I am fine with that too. Since remainingAmount stores amount that can be withdrawn at any given point of time (assuming sufficient balance), so with this, withdrawableAmountOf can be calculated as min(balance, remainingAmount), right?

it was a question from me to you :)) - but now i realized that this won't work

no matter how much is streamed the recipient must be able to withdraw the `remainingAmount`, because this amount will always updated with the tokens that have already been streamed 

so the new invariant will be: `withdrawableAmount = min(balance, streamedAmount) + remainingAmount`

--- 

there are these cases:
1. stream **not canceled**

- remainingAmount _is zero_
  - balance <= streamedAmount: `withdrawableAmount == balance`
  - balance > streamedAmount: `withdrawableAmount == streamedAmount`
- remainingAmount _is not zero_
  - balance <= streamedAmount: `withdrawableAmount == balance + remainingAmount`
  - balance > streamedAmount: `withdrawableAmount == streamedAmount + remainingAmount`

2. stream **canceled**

- `withdrawableAmount == remainingAmount`

correct me if i missed smth

---

> what if the newAmountAdded is greater than balance? you should be able to withdraw this amount, but since the balance is less, would it revert? 
I didn't get this. Could you please explain again? 

i was trying to point out edge case: `remainingAmount > balance && streamedAmount < balance`

> Yes. withdraw can be called even after it has been cancelled. Since cancel would not transfer funds to the recipient anymore. Another advantage of this is that if the recipients get blocked by stablecoins such as USDT / USDC (may be because they exploited another protocol), sender would be able to cancel and refund his remaining amount. While in the current design, sender funds can get locked.

yes, this whole discussion started due to this issue :))

but in my [initial message](https://github.com/sablier-labs/v2-open-ended/discussions/34#discussioncomment-9156560) i said that we could add a new function `withdrawWhenCanceled`, and then, you suggested to add a new amount instead which results in making `withdraw` callable when canceled

---

> It depends on the time variable passed in withdraw function. if its block.timestamp it should be min(remainingAmount, balance).
what time variable do you pass in the withdraw function?

i don't think you got my point (sorry for not being super clear)

consider this example:
- a stream is created (sufficient balance) with a `rps` that generates 1000 tokens per day  `t0` 
- after exact 12 hours ( `t1` ) the sender adjusts the `rps` to 2000 tokens per day
   - the `remainingAmount` is updated with 500
   - `lastTimeUpdate` is updated to `t1`
- after 1 more second (`t1 + 1` ) the recipient wants to withdraw less 500

in the current design this won't be possible

> Its an input variable.

yeah

---

will start working on this asap, and it will be easier once we get things moving

**Reply by andreivladbrg** [reply]
> Why are you pessimistic about making it work?
if we don't have something stable for the lockup version - what do you mean by this?

sorry for not being clear, i was referring to the NFT marketplace use case: yes, we will offer this NFT feature, but i am just pessimistic that it will "work" (by work i mean that if it will be actually used) because it would be too complex to make the system work with an open-ended contract. since with the lockup, where it is less complex, we don't have an integration made already

**Reply by smol-ninja** [reply]
I've not reviewed the full comment yet but the gas comparison is not correct. Since during `create`, `remainingAmount` is the only field which will still be 0, we should put it in slot 3 and not in slot 1. Then, the gas consumption will be nearly same. I ran the tests with the following struct, and its 137,586 i.e. only 2.5k higher.

This is because `create` will write into slot 0, 1 and 2 but not 3.

```solidity
struct StreamWithAmount {
    // slot 0
    uint128 balance;
    uint128 ratePerSecond;
    // slot 1
    address sender;
    uint40 lastTimeUpdate;
    bool isStream;
    bool isCanceled;
    // slot 2
    address asset;
    uint8 assetDecimals;
    // slot 3
    uint128 remainingAmount;
}
```

**Reply by andreivladbrg** [reply]
> we should put it in slot 3 and not in slot 1. Then, the gas consumption will be nearly same

correct, didn't know we could do that, thanks

**Reply by smol-ninja** [reply]
> so the new invariant will be: `withdrawableAmount = min(balance, streamedAmount) + remainingAmount`

I think there is a misunderstanding here. When I mentioned `remainingAmount`, I meant to use it in a different context.

In the current design, the calculation for the withdrawable amount depends on `lastTimeUpdate`. So when a stream is cancelled, we update its value and trigger the transfers. If we want to avoid withdrawing during `cancel`, we cannot update `lastTimeUpdate`. And as a result, the future calculation of `_streamedAmountOf` could be incorrect. 

So my suggestion was that the `remainingAmount` would represent the available amount to withdraw, given there is sufficient balance. That is, it will include `_streamedAmountOf` + amounts from the previous checkpoints, allowing us to safely update `lastTimeUpdate` independently (for example, when the user restarts the stream).

### An example
- When a user withdraws, `remainingAmount += _streamedAmountOf`. `_streamedAmountOf` also updated the `lastTimeUpdate`. Then it will take the `min(remainingAmount, balance)`.
- When a user cancels a stream, it will update `remainingAmount` the same as above. In the future, when a user requests to withdraw on a canceled stream, we can just use `min(balance, remainingAmount)` without looking at / updating `lastTimeUpdate`.

Let me know if I have not made myself clearer or if this is less better than the one you had in your mind.

**Reply by andreivladbrg** [reply]
i see that we are not understanding each other :))

>  If we want to avoid withdrawing during cancel, we cannot update lastTimeUpdate. And as a result, the future calculation of _streamedAmountOf could be incorrect

if the stream is canceled, the streamed amount will always be zero (because we set `rps = 0` - doesn't matter what value is `lastTimeUpdate`). the cancel function, instead of transferring the funds to the recipient, will simply sum the `remainingAmount += min(remainingAmount, balance)`, and as said in my previous message:

<img width="342" alt="image" src="https://github.com/sablier-labs/v2-open-ended/assets/99738872/cf33ce82-c167-4e23-ad2c-5fa7b3193bb3">

--- 

i was not focusing on the canceled streams, as it is simply how to handle that, but on the withdraw function when remainingAmount is not zero.

the `remainingAmount` will only be updated with `min(balance, remainingAmount)`, and the functions that will update this variable are: `adjustRatePerSecond`, `cancel`, and `refundFromStream`

one other aspect is that `lastTimeUpdate` will also be updated in these functions

> An example

i will proceed with an implementation - it will be easier to see the code

**Comment by andreivladbrg** [comment]
I will tentatively close this discussion, as we have softly agreed, and I will reopen it in case of implementation issues.
